grakel_replace/torch_wl_kernel.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import annotations
+
+from collections import Counter
+
+import networkx as nx
+import torch
+from torch import nn
+
+
+class TorchWLKernel(nn.Module):
+    """Custom PyTorch implementation of Weisfeiler-Lehman Kernel.
+
+    Args:
+        n_iter: Number of WL iterations
+        normalize: Whether to normalize the kernel matrix
+    """
+
+    def __init__(self, n_iter: int = 5, normalize: bool = True):
+        super().__init__()
+        self.n_iter = n_iter
+        self.normalize = normalize
+        self.label_dict = {}
+        self.label_counter = 0
+
+    def _get_sparse_adj(self, graph: nx.Graph) -> torch.sparse.Tensor:
+        """Convert NetworkX graph to sparse adjacency tensor."""
+        edges = list(graph.edges())
+        if not edges:
+            num_nodes = graph.number_of_nodes()
+            return torch.sparse_coo_tensor(
+                indices=torch.empty((2, 0), dtype=torch.long),
+                values=torch.empty(0),
+                size=(num_nodes, num_nodes),
+                device=self.device
+            )
+
+        # Create COO format indices
+        row = torch.tensor([e[0] for e in edges], dtype=torch.long)
+        col = torch.tensor([e[1] for e in edges], dtype=torch.long)
+        edges = torch.stack([
+            torch.cat([row, col]),  # Add both directions for undirected graph
+            torch.cat([col, row])
+        ])
+
+        values = torch.ones(edges.size(1), dtype=torch.float)
+        N = graph.number_of_nodes()
+
+        return torch.sparse_coo_tensor(
+            edges, values, (N, N),
+            device=self.device
+        )
+
+    def _init_node_labels(self, graph: nx.Graph) -> torch.Tensor:
+        """Initialize node label tensor from graph."""
+        # Get node labels and convert to indices
+        labels = []
+        for node in range(graph.number_of_nodes()):
+            label = graph.nodes[node].get("label", str(node))
```

Comment:
When debugging things, this `get()` doesn't actually seem to work, i.e. I get a `KeyError` instead of it returning the default of `str(node)` if it doesn't exist.

Not your fault but whatever `graph.nodes[node]` is, doesn't seem to implement `get` correctly.